DietPi-Software | TasmoAdmin

svh1985 opened the pull request:
**Formal software information**
- Software title | TasmoAdmin
- Short description | TasmoAdmin is an administrative Website for Devices flashed with Sonoff-Tasmota.

**Status**: Ready
- [x] Install
- [x] Uninstall 
- [x] Apache config
- [x] lighttpd config
- [x] nginx config
- [x] create and change forum reference
- [x] Changelog entry

**Reference**: https://github.com/MichaIng/DietPi/issues/3103

svh1985 committed:
Added TasmoAdmin to list, Install and Uninstall functions

svh1985 commented:
Note:

During install get error:
`/DietPi/dietpi/dietpi-software: line 2185: ((: !  : syntax error: operand expected (error token is "!  ")`

MichaIng commented:
@svh1985 
Many thanks for doing this.

> `/DietPi/dietpi/dietpi-software: line 2185: ((: ! : syntax error: operand expected (error token is "! ")`

It's the line `if (( ! $G_INTERACTIVE )); then`. This variable should always be created by https://github.com/svh1985/DietPi/blob/dev/dietpi/func/dietpi-globals#L37-L50, although it was newly added as replacement for `G_USER_INPUTS`, which IMO was not too clear. Did you assure that all scripts on your testing system are on your current dev branch?
However the error has no further effect besides having all installs interactive.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -6547,6 +6560,21 @@ Mycroft AI requires at least 2 GiB memory on first start. We will now increase y
+
+			Download_Install 'https://github.com/reloxx13/TasmoAdmin/archive/master.zip'
+			cp -a TasmoAdmin-master/tasmoadmin/ /var/www/tasmoadmin/
+			rm -R TasmoAdmin-master
```
@svh1985 
What you thin about `mv TasmoAdmin-master/tasmoadmin /var/www/tasmoadmin`?

svh1985 replied:
Sounds good! I looked at other implementations and copied that :P

MichaIng replied:
@svh1985 
Ah what I did not think about:
- `mv TasmoAdmin-master/tasmoadmin /var/www/tasmoadmin` will fail if the target dir already exists, thus on reinstalls it will not be touched.
- `cp -a TasmoAdmin-master/tasmoadmin/ /var/www/tasmoadmin/` merges the dir inside, although currently not sure if this leads to `/var/www/tasmoadmin/tasmoadmin` if target dir exists.
**EDIT:**
```
2019-10-06 17:47:42 root@micha:/tmp# mkdir source
2019-10-06 17:47:50 root@micha:/tmp# cp -a source target
2019-10-06 17:47:58 root@micha:/tmp# cp -a source/ target/
2019-10-06 17:48:06 root@micha:/tmp# l target/
total 0
drwxr-xr-x 2 root root 40 Oct  6 17:47 source
```
- So to assure the new dir is merged into the existing one, ist must be:
```
mkdir -p /var/www/tasmoadmin
cp -a TasmoAdmin-master/tasmoadmin/. /var/www/tasmoadmin/
```
- The dot `.` matches all (also hidden) included files, but not the parent dir.

Actually to assure clean reinstalls:
- Does TasmoAdmin have an internal (web UI or CLI-based) updater? In this case we should simply skip downloading and installing `/var/www/tasmoadmin` if this dir already exists. So reinstalls will re-apply configs, install pre-reqs, PHP and webserver modules, but the PHP scripts themselves stay untouched for internal updater, which e.g. might include some required migration steps and such.
- If no internal updater exist, I guess `/var/www/tasmoadmin` contains some config file or other files/dirs that are expected to be changed/customised by user? We could then merge those files from existing dir to fresh download, remove target dir and move fresh download with custom file/dir merged into place.
- If all custom/settings are saved outside of `/var/www/tasmoadmin`, we could completely remove existing dir and move the new in place.

svh1985 replied:
> @svh1985
> * Does TasmoAdmin have an internal (web UI or CLI-based) updater? In this case we should simply skip downloading and installing `/var/www/tasmoadmin` if this dir already exists. So reinstalls will re-apply configs, install pre-reqs, PHP and webserver modules, but the PHP scripts themselves stay untouched for internal updater, which e.g. might include some required migration steps and such.

There is a way to update via the web page itself, I can check if we can also use that method from the command line.
![SelfUpdate_-_TasmoAdmin](https://user-images.githubusercontent.com/5213210/66331540-1cf60580-e933-11e9-942d-ecb675597baa.png)

> * If no internal updater exist, I guess `/var/www/tasmoadmin` contains some config file or other files/dirs that are expected to be changed/customised by user? We could then merge those files from existing dir to fresh download, remove target dir and move fresh download with custom file/dir merged into place.


> * If all custom/settings are saved outside of `/var/www/tasmoadmin`, we could completely remove existing dir and move the new in place.

All relevant data is stored in : `/var/www/tasmoadmin/data`, so I think this should be backuped somehow.

MichaIng replied:
@svh1985
In this case, IMO it is the best solution to skip download if `/var/www/tasmoadmin` exist and instead prompt an info that user should use the internal updater. All other things like pre-reqs and required webserver drop-in configs should still be applied, so it is only about the web app scripts themselves.

Check how we do this with Pydio as an example: https://github.com/MichaIng/DietPi/blob/master/dietpi/dietpi-software#L5078-L5092

This is simply safest, e.g. if a newer version does not work with old configs, e.g. some migration steps required that are done by internal updater then. Also if some database is used, this is often the case.

If there is some CLI available to call the internal updater, this would be an alternative, but depends as well on if e.g. some (optionally) interactive steps are done that would need to be automated on non-interactive installs, respectively this software title then excluded for non-interactive dietpi-software calls.

Good to know about the config/data dir regardless. Long term I'd like to have some software backup feature that creates backups for certain software configs and data, also to allow easier migration of systems to new SBCs/images.

svh1985 replied:
I've added that part, but for some reason the name of the software is not shown:
![DietPi_VirtualBox](https://user-images.githubusercontent.com/5213210/66414560-f0f18780-e9f9-11e9-8138-e64562d29542.png)

```
G_DIETPI-NOTIFY 2 "**${aSOFTWARE_WHIP_NAME[$software_id]}** install dir \"/var/www/pydio\" already exists. Download and install steps will be skipped.
 - Please manually backup your config files+data, remove the install dir and rerun \"dietpi-software (re)install $software_id\" if you need to reinstall.
 - If you want to update the **${aSOFTWARE_WHIP_NAME[$software_id]}** instance, please use the internal updater from WebUI."
```

MichaIng replied:
@svh1985
Ah this is due to recent change, I renamed the `aSOFTWARE_WHIP_NAME` to `aSOFTWARE_NAME`. Is the first still used/referenced somewhere in current code? At least the wiki has already been updated.

svh1985 replied:
Okay, thanks for checking 😄 I'll change it.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -6547,6 +6560,21 @@ Mycroft AI requires at least 2 GiB memory on first start. We will now increase y
+			rm -R TasmoAdmin-master
+			chown -R www-data:www-data /var/www/tasmoadmin
+
+			[[ $DEPS_LIST ]] && G_DIETPI-NOTIFY 2 'Installing required PHP modules' && G_AGI $DEPS_LIST
```
`DEPS_LIST` packages, when assigned, are installed automatically by the `Download_Install` function, parallel to the archive download, so no need to do this manually.

svh1985 replied:
So if I understand correctly I can change it to:
`G_DIETPI-NOTIFY 2 'Installing required PHP modules'`

MichaIng replied:
@svh1985
Correct would be "Installed ..." since this was already done when this notify call reached :wink:. Or you move the notification some lines up that it is printed before Download_Install call installs those packages.

Actually I simply left those info prints out when done via Download_Install. G_AGI is IMO verbose enough about what is done?
_________
You can also install the packages manually first (and not apply to DEPS_LIST variable), but the intention with Download_Install() was to run installs and downloads in parallel to speed up things a bid. The download runs in background then (G_THREAD_START) while the APT installs run in foreground with usual output.

svh1985 replied:
I've removed line 6574.
What does G_AGI do? Do I need to let it there?

MichaIng replied:
@svh1985 
G_AGI == non-interactive apt-get install, so when Download_Install is skipped, it should be called instead.
`[[ $DEPS_LIST ]] && G_DIETPI-NOTIFY 2 'Installing required PHP modules' && G_AGI $DEPS_LIST` can then be removed.

svh1985 replied:
Okay clear, I've removed it!

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1837,6 +1837,19 @@ DietPi-Software will decrypt and use it for software installs. You can change it
 		aSOFTWARE_ONLINEDOC_URL[$software_id]='p=70#p70'
 
+		#------------------
+		software_id=177
```
Software IDs 27 and 140 are now available as well, so no need to increase the maximum number currently 😃.
```
root@VM-Buster:~# dietpi-software free
[  OK  ] DietPi-Software | Initialised database
[  OK  ] DietPi-Software | Reading database
Free software ID(s): 27 140
```

svh1985 replied:
Will change it to ID 27

svh1985 commented:
@MichaIng TasmoAdmin requires mod_rewrite, what do you suggest to do;
-virtual directory with custom settings and location eg /var/tasmoadmin (better suggestions are welcome)
-place in /var/www/tasmoadmin and change the default config by enabling mod-rewrite

ps. Ive double checked the file to make sure I got the latest.

> @svh1985
> Many thanks for doing this.
> 
> > `/DietPi/dietpi/dietpi-software: line 2185: ((: ! : syntax error: operand expected (error token is "! ")`
> 
> It's the line `if (( ! $G_INTERACTIVE )); then`. This variable should always be created by https://github.com/svh1985/DietPi/blob/dev/dietpi/func/dietpi-globals#L37-L50, although it was newly added as replacement for `G_USER_INPUTS`, which IMO was not too clear. Did you assure that all scripts on your testing system are on your current dev branch?
> However the error has no further effect besides having all installs interactive.

MichaIng commented:
Ref: https://github.com/MichaIng/DietPi/issues/3103#issuecomment-538755183

svh1985 committed:
Update dietpi-software

-Changed software_id to 27
-Changed install procedure with improvements from MichaIng

svh1985 committed:
Create apache.tasmoadmin.conf

Added Apache Directory conf.

svh1985 committed:
Added checks for difference webservers

svh1985 committed:
Merge branch 'dev' of https://github.com/svh1985/DietPi into dev

svh1985 commented:
> @svh1985
> Are those options really required?
> 
> * `Indexes` lists the raw dir content, if URL does not point to explicit file and no index file was found. This should be not set by default since it potentially reveals information that this at best not required for clients/visitors and is an information leak. It only makes sense for public download dirs where visitors might want to browse available downloads.
> * `FollowSymLinks` should be only added if there are symlinks expected from and to inside this dir structure and otherwise no web application should link to any place outside of its own subdir.
> * `ExecCGI` not required since it requires `mod_cgi` but we use `mod_php`.
> * `Includes` allow syntax like `<!--#element attribute=value attribute=value ... -->` with `mod_include`, but since neither the module is installed/enabled nor I see such syntax in TasmoAdmin .htaccess/scripts, it should be not required.
> 
> Note that the less options used, the faster requests are handled, `Options None` should work and then be used?
> 
> Great work so far ❤️, especially how well you master the coding style. If you face any difficulties or anything missing or unclear in [Wiki](https://github.com/MichaIng/DietPi/wiki/How-to-add-a-new-software-title), feel free to mention and we can add/enhance this.

Options None did not work. Leaving out Options entirely did.

MichaIng commented:
@svh1985 
Okay, then either Indexes or FollowSymLinks was required, since those two are enabled by default on /var/www + subdirs.

svh1985 committed:
Apache and install

Added installed check
Changed apache Directory conf

svh1985 committed:
Option FollowSymLinks

Added Option FollowSymLinks as this is required to run TasmoAdmin.

svh1985 committed:
Update dietpi-software

Removed notify and moved webserver checks into the already installed check.

svh1985 commented:
> @svh1985
> Okay, then either Indexes or FollowSymLinks was required, since those two are enabled by default on /var/www + subdirs.

I've checked and the FollowSymLinks were required. Added that back to the config in https://github.com/MichaIng/DietPi/pull/3143/commits/aa56952a147d84a5910491bd0df67a7588889b77

svh1985 committed:
Create lighttpd.tasmoadmin.conf

Finished lighttpd config

svh1985 committed:
Update dietpi-software

Changes:
-Changed aSOFTWARE_WHIP_NAME to aSOFTWARE_NAME function
-Added the webserver checks and download/install of configuration files.

svh1985 committed:
Create nginx.tasmoadmin.conf

First try at the nginx config. Does not work yet.

svh1985 commented:
@MichaIng I'm getting the hang of it 🤓. I've added the config for the different web servers, although Nginx is still not working, so I think I'm gonna need your help for that. lighthttpd worked here, but maybe you want to double check as well. One important thing to note is that the /var/www/tasmoadmin/data folder has two files that may not be downloaded directly and should be blocked: **devices.csv** and **MyConfig.json**, they both contain passwords.

I can't test the `dps_index=$software_id Download_Install 'lighttpd.tasmoadmin.conf' $tasmoadmin_conf` mechanism,  because the configs are not yet merged on github.

MichaIng commented on .conf/dps_27/nginx.tasmoadmin.conf:
```diff
@@ -0,0 +1,42 @@
+		rewrite ^/([a-zA-Z_]+)/([a-zA-Z_]+)/?$ /index.php?page=$1&action=$2;
+		rewrite ^/([a-zA-Z_]+)/([0-9]+)/?$ /index.php?page=$1&device_id=$2;
+		rewrite ^/([a-zA-Z_]+)/?$ /index.php?page=$1;
+	}
```
@svh1985 
- As you are already inside /tasmoadmin location directive, the above second one is obsolete, as it will match in every case? Or is this to prevent even parsing the rewrite rules if one of the above locations is matched?
- The `/tasmoadmin` subdir needs to be inside the rewrite origin and target, like it is in the Lighttpd config above.
- I am not 100% sure how things are called within Nginx, but AFAIK all the rewrite rules need to be in front of the *.php location directive since it's not called anymore otherwise, at least for the rewrite rules without `last` flag? But should be simply tested. If it indeed works or is even required, then you could replace "last" by "break" which AFAIK means that no further rewrite rules are parsed, which just makes sense in those cases.
________
Btw just found that Apache `FollowSynLinks` is required for rewrites to work: https://httpd.apache.org/docs/current/mod/mod_rewrite.html#rewriterule
Didn't know that 🙂.

MichaIng commented on .conf/dps_27/nginx.tasmoadmin.conf:
```diff
@@ -0,0 +1,42 @@
+	}
+
+	# Enable URL rewrites, required for sync clients
+	try_files $uri $uri/ /pydio/index.php?$args;
```
@svh1985 
This needs to be removed as it would override the below rewrite rules as well?

MichaIng commented:
@svh1985 
> I can't test the dps_index=$software_id Download_Install 'lighttpd.tasmoadmin.conf' $tasmoadmin_conf mechanism, because the configs are not yet merged on github.

You should be able to test it when switching to your branch:
```
G_CONFIG_INJECT 'DEV_GITOWNER=' 'DEV_GITOWNER=svh1985' /DietPi/dietpi.txt
G_CONFIG_INJECT 'DEV_GITBRANCH=' 'DEV_GITBRANCH=dev' /DietPi/dietpi.txt
dietpi-update -1
```

However need to do extra effort just to test this function 😉.

svh1985 commented:
@MichaIng Thanks for the tip! Does this mean that if I commit something it will also get the latest commits when updating?

The dps doesn't work tho, still looks at the original branch:
![DietPi_VirtualBox](https://user-images.githubusercontent.com/5213210/67638063-27c40a80-f8e1-11e9-9337-031805f5616a.png)

MichaIng commented:
@svh1985 
Indeed, that does not make sense, should be the applied Git owner. Just fixed it: https://github.com/MichaIng/DietPi/commit/6af0e9b0233f6b88e575c8e2d21f1aa7013be8db

svh1985 committed:
Merge pull request #2 from MichaIng/dev

Dev

svh1985 committed:
Update nginx.tasmoadmin.conf

Changed nginx config. Works for me, please test.

svh1985 commented:
@MichaIng that definitely helped a lot! 😄
It took some time.. but I got the nginx to work! 🎉

I guess the form post should also be created and then I'll add the ID's to the code as ref. Is creating the software post on the forum something I can do or should you do that?

MichaIng commented:
@svh1985 
Many thanks for your work on this 👍! Will run some tests tomorrow.

svh1985 commented:
@MichaIng 
https://github.com/reloxx13/TasmoAdmin/issues/307 Update TasmoAdmin from CLI 
Got a response; suggestion to use git to get the files and also use git to update. The issue got marked as an enhancement so cli update might be implemented one day.

MichaIng commented:
@svh1985 
Okay, if the suggestion is to use Git, then there are obviously no internal migration steps required. I'd skip the Git client however and do a raw fresh download by preserving config files as you suggested above, when an existing install is found (reinstall). This allows updating via dietpi-software then as well.

The Git client is nice for developers, but for end users it is simply an unnecessary overhead: You need to tell it to not download the whole repository with all branches and all commits every done. If it is simply to have the latest files, a simple master archive download is all that is required

svh1985 commented:
@MichaIng I propose we leave it as is for now, I noticed the internal web updater is also updating to beta releases, so there will always be updates after install. Hopefully the CLI updater will be implemented soon. :)

Let me know if you found anything while testing

MichaIng commented:
#### Testing
🈴 Install: Webserver and PHP must be installed before TasmoAdmin, otherwise e.g. modules cannot be enabled, potentially wrong PHP handler is pulled etc. **Fixed: https://github.com/MichaIng/DietPi/pull/3143/commits/0af99b35456cf7aff15c12aac21102ea113a2106**
🈯️ Lighttpd: Web UI, SelfUpdate, .htaccess + data access block
🈯️ Apache
🈯️ Nginx

MichaIng committed:
v6.27

+ DietPi-Software | TasmoAdmin: Minor alignment

MichaIng committed:
v6.27

+ DietPi-Software | TasmoAdmin: Minor alignment

MichaIng committed:
v6.27

+ DietPi-Software | TasmoAdmin: Deny access to .htaccess and dot files in general, stop search after "/tasmoadmin" location matches, simplify assets regex and make it case-insensitive + minor alignment

MichaIng committed:
v6.27

+ DietPi-Software | TasmoAdmin: Install after webservers + PHP and uninstall before webserver + PHP (which should be done with all web applications)
+ DietPi-Software | TasmoAdmin: Minor alignment

MichaIng commented:
@svh1985 
Works well on all webserver during my tests, although I could not test some actual modules. Great work 👍! Reordered install that it is done after webserver and PHP and did some code alignments.

Will create an online docks page later, not much to do since the install and first login simply does all that is needed. Free first user + password registration is best anyway.

If you have some nice explanation and/or screenshot, I will be happy to add it, otherwise will go simply with what I find on thei GitHub/website 😉.
**Started with online docs: https://dietpi.com/phpbb/viewtopic.php?p=20584#p20584**

And if you could add some CHANGELOG.txt entry, this would be perfect, otherwise I can add it with a separate commit as well.
___________
Adjusted required Apache modules to match what is used within the `.htaccess` files: https://github.com/MichaIng/DietPi/pull/3143/commits/e32ac6a0aa695db38757eef37de83c354455e3b5
__________
Okay will merge now. Fine tuning can still be done at a later time.

MichaIng committed:
v6.27

+ DietPi-Software | TasmoAdmin: Add online docs URL

MichaIng committed:
v6.27

+ DietPi-Software | TasmoAdmin: Adjust required Apache modules:
    - /tasmoadmin .htaccess file: https://github.com/reloxx13/TasmoAdmin/blob/master/tasmoadmin/.htaccess, 
    - /tasmoadmin/data .htaccess file: https://github.com/reloxx13/TasmoAdmin/blob/master/tasmoadmin/data/.htaccess
    - SetEnvIfNoCase requires mod_setenvif: https://httpd.apache.org/docs/current/mod/mod_setenvif.html#setenvifnocase
    - Rewrite rules require mod_rewrite: https://httpd.apache.org/docs/current/mod/mod_rewrite.html
    - "Require" requires mod_authz_core: https://httpd.apache.org/docs/current/mod/mod_authz_core.html#require
    - AuthType and AuthName require mod_authn_core: https://httpd.apache.org/docs/current/mod/mod_authn_core.html#authname
    - AuthUserFile requires mod_authn_file: https://httpd.apache.org/docs/current/mod/mod_authn_file.html#authuserfile
    - mod_headers, mod_env, mod_dir and mod_mime are not explicitly required.

svh1985 commented:
@MichaIng I was thinking, maybe we should give the app a specific port to increase the security. If you run a site and map the port in your router you also expose TasmoAdmin. What do you think?

MichaIng commented:
@svh1985
Good point.

An own port is not trivial since it uses a regular web server.

If it has no internal login mask, we should add a webserver-based authentication. I'll check and add this.

#### EDIT: It has a login mask, do you think it's still too risky? We could block access from any host/ip that is not from local network, like we do for Pi-hole, although some users might want to configure remotely, hence we should do this optionally?
